### ISSUE TITLE
Throw instead of return false on mistyped matcher arguments

### ISSUE THREAD

**Issue by joshua-i** [dart-lang/matcher#48]
Some Matchers typecheck for specific types that they know how to handle. Currently, they return false if it is not one of the expected types. This leads to undesirable behavior. For example, the following passes:
`expect(123, isNot(contains(2));`

It would be better for contains to throw when it receives anything other than an expected type. From what I can find, the other matchers that have this behavior are
- everyElement
- everything in string_matchers.dart
- everything in numeric_matchers.dart

**Comment by joshua-i** [comment]
On closer inspection, between string_matchers and numeric_matchers, the only one that could actually receive an unexpected type is string_matcher's matches. The others just do type-checking as a precaution, I guess, though it wouldn't hurt to throw there, too.

**Comment by joshua-i** [comment]
Nope, confused myself. It is a problem for everything in string_matchers and numeric_matchers.

**Comment by nex3** [comment]
This is tricky. We get real value from doing those checks the way we do now—the error reporting is much better for complex matchers if you know exactly where the invalid value comes from. At the same time, your example definitely doesn't have the expected behavior.

I think this is something to deal with in the Grand Matcher Refactor. We'll want to give matchers the ability to return a failing result that's tagged as an error for consumers where that's relevant.

**Comment by joshua-i** [comment]
Can you give an example of the complex matcher case you're describing, where returning false is better than throwing? Not sure I understand.

**Comment by nex3** [comment]
You might write

```dart
expect(myObject.future, completion(
    contains(anyOf([
      isNull,
      containsPair(contains("foo"), isGreaterThan(3))
    ]))));
```

If one of the map keys isn't a string, you want it to tell you that the future completed to an iterable that contained neither `null` nor a map with a key that contains "foo". You don't want it to just throw an error that says "12 isn't a string" without any further context.

**Comment by joshua-i** [comment]
Is that more important than tests potentially passing incorrectly? And the errors/confusion you get when these silently fail to match when used in Mockito behavior-setting are even worse.

**Comment by nex3** [comment]
Regardless of anything else, throwing an error would be a breaking change. I don't intend to release a breaking version of matcher in the short term, and once we get to the long term we might as well do tagging.

Also, throwing an error *also* produces incorrect results. In the example I gave above, it's fine if the iterable contains a map with an integer argument as long as it also contains a `null` value or a matching map. If `contains()` threw in that circumstance, that would be incorrect.
